### ISSUE TITLE
Nullable state not tracked through switch expression branch

### ISSUE THREAD

**Issue by jaredpar** [dotnet/roslyn#48861]
The following code incorrectly warns on the `o.ToString` call

```cs
#nullable enable

string M(object o) => (o is not null) switch 
{
  true => o.ToString(),
  false => ""
};
```

**Comment by mattzink** [comment]
Dumb question: Isn't `(o is not null)` always true since o is not nullable?

**Comment by jaredpar** [comment]
@mattzink yes the annotation specifies this is non-null but it's still possible for `null` values to be passed. Couple of scenarios outlined below:

```cs
// Code that existed before C# 8 or after C# but has nullability disabled
M(null);

// Code in enabled nullable contexts. This is bad but possible 
M(null!);
```

The nullable annotations are not a guarantee of non-null. Rather there a strong hint to the compiler and developer where `null` can and cannot be used. It's a system very similar to how TypeScript handled nullable values.

**Comment by mattzink** [comment]
@jaredpar Thanks for reminding me the rest of the world hasn't just jumped onto the nullable-reference-types-are-gospel bandwagon like my team has. We don't have to worry about pre-C# 8 callers and all uses of `!` are heavily scrutinized during code reviews.

**Comment by SunnieShine** [comment]
In fact... Not only in switch expression, but in switch block...
```csharp
private static bool SetEquals(Set<T>? left, Set<T>? right)
{
    switch ((left, right))
    {
        case (null, null):
        {
            return true;
        }
        case (not null, not null):
        {
            foreach (var element in left._list) // Here.
            {
                if (!right._list.Contains(element)) // Here.
                {
                    return false;
                }
            }

            return true;
        }
        default:
        {
            return false;
        }
    }
}
```

![](https://ftp.bmp.ovh/imgs/2020/10/0a4546d2a7e39094.png)

:(

**Comment by RikkiGibson** [comment]
This particular scenario appears to have been fixed. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgYhgVwDaosVBTABNmTHAWACgyBZAChmwHdcB7YAK2wGMAXKgSh4G4yZAAIBGAAwB+XNWZsuTHrgC8APlxVGuKAGdcMRp31pUSnXSid2AC1xkA3mVy5OAJ2R41TAHQAVRgDKbrAA5rwANE64AGYQqDqe6gBESWQAvvxAA===)
